[border-agent] allow setting UDP port at runtime

milas opened the pull request:
~Make `Start(aUdpPort)` public so that it's possible to explicitly choose the UDP port at runtime rather than only at compile-time.~

EDIT: See latest revision - a `SetUdpPort` setter is exposed instead of making the start method w/ port public per review feedback.

This retains existing compatibility and support for the compile-time `OPENTHREAD_CONFIG_BORDER_AGENT_UDP_PORT` as the "default". Frankly, I'm not sure there's any value in its existence aside from backwards compatibility.

I did change the comment to describe the `0` value for the port as being "dynamically-assigned", which is e.g. how IANA refers to this and avoids confusion with "ephemeral mode" on the agent itself.

My plan for this is to add a `--border-agent-port` CLI flag or similar to `ot-br-posix` in a future Pull Request.

See also:
 * https://github.com/orgs/openthread/discussions/11093

abtink commented on src/core/meshcop/border_agent.hpp:
```diff
@@ -146,9 +151,20 @@ class BorderAgent : public InstanceLocator, private NonCopyable
+     * @retval kErrorNone  Successfully started the Border Agent.
+     * @retval ...         Failed to start the Border Agent (e.g., could not bind to UDP port).
+     */
+    Error Start(uint16_t aUdpPort);
```
Thanks, @milas.

For the use case of allowing the port number to be specified, we don't need to make the `Start()` method `public`. The Border Agent itself will start/stop based on events (e.g., when attached/detached, etc.).

Note that #11087 is making `Start()` and `Stop()` private. They were exposed as `public` to allow a workaround solution due to sharing the underlying DTLS `SecureAgent` between multiple modules. The `Commissioner` module would stop and restart the `BorderAgent` when it was itself enabled/disabled. This was more of a hack to allow certain tests to be performed when the two features were used together. The PR above is removing this hack.

Regarding your use case, here is my suggestion:

- We can add a new method, `Error SetUdpPort(uint16_t aUdpPort)`, which sets the default UDP port to be used by BA. We should restrict this to when BA is stopped (return an error otherwise).
- We track the default in a new member variable, `mDefaultUdpPort`, which we initialize with `kUdpPort` from the `OPENTHREAD_CONFIG_BORDER_AGENT_UDP_PORT`.
- This new API can then be exposed in `openthread/border_agent.h` as well (as `otBorderAgentSetUdpPort(...)`).
- I would also suggest adding a corresponding CLI command for this (e.g., update `@cli ba port` to allow both get and set).

milas commented:
Thanks for the quick review & very clear feedback - I'm moving this to draft for now until I can revise the approach based on the review

milas committed:
[border-agent] allow choosing port at runtime

Add a setter for the UDP port and expose it via the C API.

Existing semantics remain around the default compile-time port, so
this change is backwards compatible.

The port for a service can only be changed while in the stopped state.

milas commented:
Revised based on feedback.

I believe things are wired up correctly.

Following `README_COMMISSIONING.md`, I do get an error:
```bash
> dataset init new
Done
> dataset commit active
Done
> ba port 12345
Done
> ifconfig up
Done
> thread start
Done
> commissioner start
Error 13: InvalidState
```

but it does appears the Border Agent comes up as expected:
```bash
> ba state
Started
Done
> ba port
12345
```

abtink commented on src/cli/cli.cpp:
```diff
@@ -412,18 +412,35 @@ template <> otError Interpreter::Process<Cmd("ba")>(Arg aArgs[])
     if (aArgs[0] == "port")
     {
-        OutputLine("%hu", otBorderAgentGetUdpPort(GetInstancePtr()));
+        if (aArgs[1].IsEmpty())
+        {
+            OutputLine("%hu", otBorderAgentGetUdpPort(GetInstancePtr()));
+        }
+        else
+        {
+            uint16_t port;
+            SuccessOrExit(error = aArgs[1].ParseAsUint16(port));
+            error = otBorderAgentSetUdpPort(GetInstancePtr(), port);
+        }
```
I think we can simplify and replace this while block with:
```c++
   return ProcessGetSet(aArgs, otBorderAgentGetUdpPort, otBorderAgentSetUdpPort);
```

abtink commented on src/core/meshcop/border_agent.cpp:
```diff
@@ -181,7 +182,22 @@ void BorderAgent::Disconnect(void)
+    return mDtlsTransport.GetUdpPort();
+}
+
+Error BorderAgent::SetUdpPort(const uint16_t &aUdpPort)
```
Better to pass as `(uint16_t aUdpPort)` instead of `const uint16_t &` (since it is simple/native type).
```suggestion
Error BorderAgent::SetUdpPort(uint16_t aUdpPort)
```

abtink commented on src/core/meshcop/border_agent.hpp:
```diff
@@ -139,16 +139,28 @@ class BorderAgent : public InstanceLocator, private NonCopyable
+     * @retval kErrorNone           Successfully set UDP port.
+     * @retval kErrorInvalidState   Service is not stopped.
+     */
+    Error SetUdpPort(const uint16_t &aUdpPort);
```
```suggestion
    Error SetUdpPort(uint16_t aUdpPort);
```

abtink commented on src/core/meshcop/border_agent.hpp:
```diff
@@ -338,6 +350,7 @@ class BorderAgent : public InstanceLocator, private NonCopyable
     Ip6::Udp::Receiver         mUdpReceiver;
     Ip6::Netif::UnicastAddress mCommissionerAloc;
     TimeoutTimer               mTimer;
+    uint16_t                   mUdpPort;
```
Suggest moving up after `State`? (we generally aim to keep member variables of similar/smaller size first).

abtink reviewed (commented):
Looks great. Thanks. 

Couple of smaller suggestions below:

jwhui commented:
@milas , are you still interested in moving this PR forward?

milas commented:
I have not had much time to spend on this, but will try to rebase it and address the feedback soon. Any maintainer is also welcome to take it over & make changes/merge if desirable -- it's a pretty trivial changeset (no expectation here ofc!)

jwhui commented:
Closing stale PR.

@milas , feel free to reopen if you would like to move this forward again.
